libmuscle/python/libmuscle/manager/deadlock_detector.py

```diff
@@ -0,0 +1,220 @@
+
+    def _process_queue_item(self, item: _QueueItem) -> None:
+        """Actually process a WAITING_FOR_RECEIVE[_DONE] request.
+
+        This method should be called inside the worker thread.
+        """
+        _logger.debug("Processing queue item: %s", item)
+        is_waiting, instance_id, peer_instance_id, port_name, slot = item
+        if is_waiting:
+            # Sanity checks, triggering this is a bug in the instance or the manager
+            if instance_id in self._waiting_instances:
+                _logger.error(
+                    "Instance %s was already waiting on a receive call. "
+                    "Did we miss a WAITING DONE event?",
+                    instance_id)
+            # Register that the instance is waiting
+            self._waiting_instances[instance_id] = peer_instance_id
+            self._waiting_instance_ports[instance_id] = (port_name, slot)
+            self._check_for_deadlock(instance_id)
+
+        else:
+            # Sanity checks, triggering these is a bug in the instance or the manager
+            if instance_id not in self._waiting_instances:
+                _logger.error(
+                    "Instance %s is not waiting on a receive call.", instance_id)
+            elif self._waiting_instances[instance_id] != peer_instance_id:
+                _logger.error(
+                    "Instance %s was waiting for %s, not for %s.",
+                    instance_id,
+                    self._waiting_instances[instance_id],
+                    peer_instance_id)
+            elif self._waiting_instance_ports[instance_id] != (port_name, slot):
+                _logger.error(
+                    "Instance %s was waiting on port[slot] %s[%s], not on %s[%s]",
+                    instance_id,
+                    *self._waiting_instance_ports[instance_id],
+                    port_name, slot)
+            else:
+                del self._waiting_instances[instance_id]
+                del self._waiting_instance_ports[instance_id]
+
+                # Check if we were part of a deadlock
+                for i, instance_list in enumerate(self._detected_deadlocks):
+                    if instance_id in instance_list:
+                        del self._detected_deadlocks[i]
+                        break
+                if not self._detected_deadlocks:
+                    # There are no deadlocks anymore: cancel shutdown
+                    self._shutdown_time = None
+
+    def _check_for_deadlock(self, instance_id: str) -> None:
+        """Check if there is a cycle of waiting instances that involves this instance.
+        """
+        deadlock_instances = [instance_id]
+        cur_instance = instance_id
+        while cur_instance in self._waiting_instances:
+            cur_instance = self._waiting_instances[cur_instance]
+            if cur_instance == instance_id:
+                self._handle_potential_deadlock(deadlock_instances)
```

Comment:
I really like the progressive delegation from `_check_for_deadlock()` to `_handle_potential_deadlock()` to `_format_deadlock()` here. Nice!